### PR TITLE
Migrate decision_* and related test from boost to catch2

### PR DESCRIPTION
**Pull request by jeffin143**
Refer https://github.com/mlpack/mlpack/issues/2523

### CODE REVIEW

File: src/mlpack/tests/decision_tree_test.cpp [hunk 1]
```diff
@@ -57,35 +55,37 @@ BOOST_AUTO_TEST_CASE(GiniGainEvenSplitTest)
...
-BOOST_AUTO_TEST_CASE(GiniGainEmptyTest)
+TEST_CASE("GiniGainEmptyTest", "[DecisionTreeTest]")
 {
   arma::rowvec weights = arma::ones<arma::rowvec>(10);
   // Test across some numbers of classes.
   arma::Row<size_t> labels;
   for (size_t c = 1; c < 10; ++c)
-    BOOST_REQUIRE_SMALL(GiniGain::Evaluate<false>(labels, c, weights), 1e-5);
+    REQUIRE(GiniGain::Evaluate<false>(labels, c, weights) ==
+        Approx(0.0).margin(1e-7));
```
Comment (zoq) [inline src/mlpack/tests/decision_tree_test.cpp:78]: Can you use `{}` for multiple lines?

File: src/mlpack/tests/decision_tree_test.cpp [hunk 2]
```diff
@@ -1033,33 +1033,33 @@ BOOST_AUTO_TEST_CASE(RandomDimensionSelectRandomTest)
   r3.Dimensions() = 100000;
   r4.Dimensions() = 100000;
 
-  BOOST_REQUIRE((r1.Begin() != r2.Begin()) ||
-                (r1.Begin() != r3.Begin()) ||
-                (r1.Begin() != r4.Begin()));
+  REQUIRE(((r1.Begin() != r2.Begin()) ||
+          (r1.Begin() != r3.Begin()) ||
+          (r1.Begin() != r4.Begin())));
```
Comment (zoq) [inline src/mlpack/tests/decision_tree_test.cpp:1038]:
```suggestion
           (r1.Begin() != r3.Begin()) ||
           (r1.Begin() != r4.Begin())));
```

File: src/mlpack/tests/decision_tree_test.cpp [hunk 3]
```diff
@@ -16,36 +16,34 @@
...
 {
   arma::rowvec weights(10, arma::fill::ones);
   arma::Row<size_t> labels;
   labels.zeros(10);
 
   // Test that it's perfect regardless of number of classes.
   for (size_t c = 1; c < 10; ++c)
-    BOOST_REQUIRE_SMALL(GiniGain::Evaluate<false>(labels, c, weights), 1e-5);
+    REQUIRE(GiniGain::Evaluate<false>(labels, c, weights) ==
+        Approx(0.0).margin(1e-7));
```
Comment (jeffin143) [inline src/mlpack/tests/decision_tree_test.cpp:39]:
```suggestion
        Approx(0.0).margin(1e-5));
```

File: src/mlpack/tests/decision_tree_test.cpp (hunk 1 above)
Comment (jeffin143) [inline src/mlpack/tests/decision_tree_test.cpp:78]:
```suggestion
    {
        REQUIRE(GiniGain::Evaluate<false>(labels, c, weights) ==
            Approx(0.0).margin(1e-5));
     }
```

File: src/mlpack/tests/decision_tree_test.cpp [hunk 4]
```diff
@@ -57,35 +55,37 @@ BOOST_AUTO_TEST_CASE(GiniGainEvenSplitTest)
...
 
   for (size_t c = 1; c < 10; ++c)
-    BOOST_REQUIRE_SMALL(GiniGain::Evaluate<true>(labels, c, weights), 1e-5);
+    REQUIRE(GiniGain::Evaluate<true>(labels, c, weights) ==
+        Approx(0.0).margin(1e-7));
 }
```
Comment (jeffin143) [inline src/mlpack/tests/decision_tree_test.cpp:83]:
```suggestion
    REQUIRE(GiniGain::Evaluate<true>(labels, c, weights) ==
        Approx(0.0).margin(1e-5));
}
```

File: src/mlpack/tests/decision_tree_test.cpp [hunk 5]
```diff
@@ -186,33 +185,33 @@ BOOST_AUTO_TEST_CASE(InformationGainEvenSplitTest)
...
   arma::Row<size_t> labels;
   arma::rowvec weights = arma::ones<arma::rowvec>(10);
   for (size_t c = 1; c < 10; ++c)
   {
-    BOOST_REQUIRE_SMALL(InformationGain::Evaluate<false>(labels, c, weights),
-        1e-5);
-    BOOST_REQUIRE_SMALL(InformationGain::Evaluate<true>(labels, c, weights),
-        1e-5);
+    REQUIRE(InformationGain::Evaluate<false>(labels, c, weights) ==
+        Approx(0.0).margin(1e-7));
```
Comment (jeffin143) [inline src/mlpack/tests/decision_tree_test.cpp:205]:
```suggestion
        Approx(0.0).margin(1e-5));
```

File: src/mlpack/tests/decision_tree_test.cpp [hunk 6]
```diff
@@ -186,33 +185,33 @@ BOOST_AUTO_TEST_CASE(InformationGainEvenSplitTest)
...
+    REQUIRE(InformationGain::Evaluate<true>(labels, c, weights) ==
+        Approx(0.0).margin(1e-7));
```
Comment (jeffin143) [inline src/mlpack/tests/decision_tree_test.cpp:207]:
```suggestion
        Approx(0.0).margin(1e-5));
```

File: src/mlpack/tests/decision_tree_test.cpp [hunk 7]
```diff
@@ -551,22 +551,22 @@ BOOST_AUTO_TEST_CASE(PerfectTrainingSet)
...
 
-    BOOST_REQUIRE_EQUAL(prediction, labels[i]);
-    BOOST_REQUIRE_EQUAL(probabilities.n_elem, 2);
+    REQUIRE(prediction == labels[i]);
+    REQUIRE(probabilities.n_elem == 2);
     for (size_t j = 0; j < 2; ++j)
     {
       if (labels[i] == j)
-        BOOST_REQUIRE_CLOSE(probabilities[j], 1.0, 1e-5);
+        REQUIRE(probabilities[j] == Approx(1.0).epsilon(1e-7));
```
Comment (jeffin143) [inline src/mlpack/tests/decision_tree_test.cpp:559]:
```suggestion
        REQUIRE(probabilities[j] == Approx(1.0).epsilon(1e-5));
```

File: src/mlpack/tests/decision_tree_test.cpp [hunk 8]
```diff
@@ -551,22 +551,22 @@ BOOST_AUTO_TEST_CASE(PerfectTrainingSet)
...
       else
-        BOOST_REQUIRE_SMALL(probabilities[j], 1e-5);
+        REQUIRE(probabilities[j] == Approx(0.0).margin(1e-7));
```
Comment (jeffin143) [inline src/mlpack/tests/decision_tree_test.cpp:561]:
```suggestion
        REQUIRE(probabilities[j] == Approx(0.0).margin(1e-5));
```

File: src/mlpack/tests/decision_tree_test.cpp [hunk 9]
```diff
@@ -594,14 +594,14 @@ BOOST_AUTO_TEST_CASE(PerfectTrainingSetWithWeight)
...
```
Comment (jeffin143) [inline src/mlpack/tests/decision_tree_test.cpp:604]:
```suggestion
        REQUIRE(probabilities[j] == Approx(0.0).margin(1e-5));
```

File: src/mlpack/tests/decision_tree_test.cpp (hunk 7 above)
Comment (jeffin143) [reply src/mlpack/tests/decision_tree_test.cpp:559]: This is not needed

File: src/mlpack/tests/decision_tree_test.cpp (hunk 1 above)
Comment (jeffin143) [reply src/mlpack/tests/decision_tree_test.cpp:78]: I will make sure of this in next test I migrate

**Pull request merged**